File: MSc-SimSci/ACM41000-Uncertainty-Quantification/assignement3/Data2LD_Fix/Data2LD_Fix/src/SmatFn.c
//
//  Last updated by Jim Ramsay 6 March 2018
//

#include <stdio.h>

void SmatFn(int    *nXbasisw, int    *nWbasisw, int    *nUbasisj, int *nAbasisj,
            int    *nrep,     double *Bvecw,    double *Avecj,    double *Ucoefj,
            double *BAtenswj, double *Smat)
{
    int nXw = *nXbasisw;
    int nWw = *nWbasisw;
    int nUj = *nUbasisj;
    int nAj = *nAbasisj;
    int nRp = *nrep;
    int nSmat = nXw*nRp;
    /* assign 0 to all positions */
    for (int p = 0; p < nSmat; p++) Smat[p]=0;
    /* loop through all columns of all four input matrices */
    for (int r = 0; r < nRp; r++)
    {
        for (int i = 0; i < nXw; i++)
        {
            for (int j = 0; j < nWw; j++)
            {
                for (int k = 0; k < nUj; k++)
                {
                    for (int l = 0; l < nAj; l++)
                    {
                        int ijkl = i*nAj*nUj*nWw + j*nAj*nUj + k*nAj + l;
                        /*  increment value at index by the product */
                        Smat[r*nXw+i] +=
                        Bvecw[j] * Avecj[l] * Ucoefj[k+r*nUj] * BAtenswj[ijkl];
                    }
                }
            }
        }
    }
}

File: MSc-SimSci/ACM41000-Uncertainty-Quantification/assignement3/Data2LD_Fix/Data2LD_Fix/R/SmatFn.R
SmatFn <- function(nXbasisw, nWbasisw, nUbasisj, nAbasisj, nrep, 
                  Bvecw, Avecj, Ucoefj, BAtens) {
  BAtens <- as.vector(BAtens)
  SvecC <- rep(0,nXbasisw*nrep)
  inprodSList <- .C("SmatFn", as.integer(nXbasisw), 
                     as.integer(nWbasisw), 
                     as.integer(nUbasisj), 
                     as.integer(nAbasisj), 
                     as.integer(nrep),
                     as.double(Bvecw), 
                     as.double(Avecj),
                     as.double(Ucoefj),
                     as.double(BAtens), 
                     as.double(SvecC))
  SmatC <- matrix(inprodSList[[10]],nXbasisw,nrep)
  return(SmatC)
}
